File: src/main/java/dev/sylvain/planning/domain/ParametresLegaux.java
package dev.sylvain.planning.domain;

import java.time.LocalTime;

/**
 * Global, admin-configurable legal parameters, loaded as a problem fact into
 * every {@link PlanningEvenement} (single instance, same pattern as
 * {@link ContrainteAdHoc}) so {@code LegalConstraints} can read them without
 * any static/global state.
 *
 * <p>The meal break travels with them although it is not a legal obligation:
 * it is the rule the organisation sets itself, judged by
 * {@code coupureRepasObligatoire} on the {@link FenetreRepas} facts projected
 * from these five fields, and used by the découpage to place the midday and
 * evening reliefs. It used to live with the découpage parameters, which is
 * where nobody looked for it, and where a grid entered as vacations never
 * read it at all (issue #438).</p>
 */
public class ParametresLegaux {

    /**
     * Absolute weekly working-time cap for every animateur, all paid. Default: 48 h = 2880 min, the ceiling set
     * by the Code du travail (art. L3121-20, durée maximale hebdomadaire
     * absolue, disposition d'ordre public) — editable from the "Constraints"
     * screen.
     *
     * <p>The Convention collective de l'Animation (ÉCLAT, IDCC 1518) used to
     * be cited beside it, at art. 5.2. Two things were wrong with that: art.
     * 5.2 is about rest days, the 48 h high week is at art. 5.7.2.3
     * (modulation) — and the organisation this deployment serves has confirmed
     * it does not fall under that convention. The Code alone founds the value.
     * See {@code docs/contraintes.md}, « La convention collective de
     * l'Animation ».</p>
     */
    public static final int DUREE_HEBDOMADAIRE_MAX_MINUTES_PAR_DEFAUT = 48 * 60;

    /**
     * Absolute weekly working-time cap for a minor ("jeune travailleur"),
     * whatever their age bracket. Default: 35 h = 2100 min, set by the Code du
     * travail art. L3162-1 (<i>« Les jeunes travailleurs ne peuvent être
     * employés à un travail effectif excédant huit heures par jour et
     * trente-cinq heures par semaine »</i>) and, for the 14-to-under-16
     * bracket employed during school holidays, by art. D4153-3 (same 35 h
     * weekly figure). Ordre public: a value above 35 h is rejected by
     * {@code ReferenceDataService.updateParametresLegaux}; a lower (more
     * protective) value stays free.
     */
    public static final int DUREE_HEBDOMADAIRE_MAX_MINEUR_MINUTES_PAR_DEFAUT = 35 * 60;

    /**
     * How long one vacation may run before the grid check says so. Default 6 h,
     * the art. L3121-16 threshold at which a break becomes legally mandatory:
     * a vacation staying strictly under it never needs an internal break, and
     * the gap before the animateur's next vacation is their pause.
     *
     * <p>It was a <em>slicing</em> setting, read from {@code ParametresDecoupage}
     * to bound what the découpage produced. The découpage is gone — every grid
     * is typed or projected from a journée type — but the rule it encoded is
     * not: it is a rule of the event, judged on whatever grid the organiser
     * ends up with, so it belongs here with the other rules, the way the meal
     * break already made that same move (issue #438).</p>
     */
    public static final int DUREE_VACATION_MAX_MINUTES_PAR_DEFAUT = 6 * 60;

    /**
     * Minimum rest required between the end of an animateur's last vacation on
     * a calendar day and the start of their first vacation the next calendar
     * day. Default: 11 h = 660 min, the absolute daily rest floor set by the
     * Code du travail (art. L3131-1). Generalizes {@code reposQuotidienMineur}
     * (night → next-day noon, minors only) to every animateur.
     */
    public static final int REPOS_QUOTIDIEN_MINIMAL_MINUTES_PAR_DEFAUT = 11 * 60;

    /**
     * How long the break that ends a working stretch lasts — <b>one duration
     * for the whole edition</b>, adult and young worker alike (ADR 0048).
     *
     * <p>Default: the thirty minutes the organisation settled on (issue #31).
     * Art. <b>L3121-16</b> owes an adult twenty at the sixth hour and that
     * twenty is the floor {@code ParametresValidator} refuses to go under; the
     * organisation gives more, because a thirty-minute relay is easier to
     * organise than a twenty-minute one and nothing in the Code forbids it.</p>
     *
     * <p>For a young worker the thirty minutes of art. <b>L3162-3</b> are
     * themselves a floor of ordre public, so {@link #dureePauseMinutes(boolean)}
     * raises a shorter edition value rather than apply it — see there. Two
     * separate fields used to carry the two brackets, which let an edition set
     * the adult one and leave the minors' one at its default, and let the
     * eligibility filter read one while the rules read the other.</p>
     */
    public static final int DUREE_PAUSE_MINUTES_PAR_DEFAUT = 30;

    /**
     * How long the meal break lasts, uninterrupted: one hour, so that a
     * two-hour window splits into two whole slots — 12-13 or 13-14 — that the
     * organiser can plan rather than endure (issue #438).
     */
    public static final int COUPURE_REPAS_MINUTES_PAR_DEFAUT = 60;

    public static final LocalTime COUPURE_REPAS_MIDI_DEBUT_PAR_DEFAUT = LocalTime.of(12, 0);
    public static final LocalTime COUPURE_REPAS_MIDI_FIN_PAR_DEFAUT = LocalTime.of(14, 0);
    public static final LocalTime COUPURE_REPAS_SOIR_DEBUT_PAR_DEFAUT = LocalTime.of(19, 0);
    public static final LocalTime COUPURE_REPAS_SOIR_FIN_PAR_DEFAUT = LocalTime.of(21, 0);

    /**
     * When the evening starts, for the equity read-out ({@code EquiteService}):
     * the minutes a poste covers past this hour are its evening hours, the
     * figure the organiser compares across the roster — « three nocturnes for
     * me, none for him » is the complaint the Équité screen exists to settle
     * before publication. A rule of the organisation, not of the law: a
     * minor's legal night stays in {@link Creneau}. Default 20:00, the earliest
     * legal night.
     */
    public static final LocalTime HEURE_DEBUT_SOIREE_PAR_DEFAUT = LocalTime.of(20, 0);

    private int dureeHebdomadaireMaxMinutes = DUREE_HEBDOMADAIRE_MAX_MINUTES_PAR_DEFAUT;
    private int dureeHebdomadaireMaxMineurMinutes = DUREE_HEBDOMADAIRE_MAX_MINEUR_MINUTES_PAR_DEFAUT;
    private int dureeVacationMaxMinutes = DUREE_VACATION_MAX_MINUTES_PAR_DEFAUT;
    private int reposQuotidienMinimalMinutes = REPOS_QUOTIDIEN_MINIMAL_MINUTES_PAR_DEFAUT;

    private int dureePauseMinutes = DUREE_PAUSE_MINUTES_PAR_DEFAUT;
    private int coupureRepasMinutes = COUPURE_REPAS_MINUTES_PAR_DEFAUT;
    private LocalTime coupureRepasMidiDebut = COUPURE_REPAS_MIDI_DEBUT_PAR_DEFAUT;
    private LocalTime coupureRepasMidiFin = COUPURE_REPAS_MIDI_FIN_PAR_DEFAUT;
    private LocalTime coupureRepasSoirDebut = COUPURE_REPAS_SOIR_DEBUT_PAR_DEFAUT;
    private LocalTime coupureRepasSoirFin = COUPURE_REPAS_SOIR_FIN_PAR_DEFAUT;
    private LocalTime heureDebutSoiree = HEURE_DEBUT_SOIREE_PAR_DEFAUT;

    public ParametresLegaux() {}

    public ParametresLegaux(int dureeHebdomadaireMaxMinutes) {
        this.dureeHebdomadaireMaxMinutes = dureeHebdomadaireMaxMinutes;
    }

    public ParametresLegaux(int dureeHebdomadaireMaxMinutes, int dureeHebdomadaireMaxMineurMinutes) {
        this.dureeHebdomadaireMaxMinutes = dureeHebdomadaireMaxMinutes;
        this.dureeHebdomadaireMaxMineurMinutes = dureeHebdomadaireMaxMineurMinutes;
    }

    public int getDureeHebdomadaireMaxMinutes() {
        return dureeHebdomadaireMaxMinutes;
    }

    public void setDureeHebdomadaireMaxMinutes(int dureeHebdomadaireMaxMinutes) {
        this.dureeHebdomadaireMaxMinutes = dureeHebdomadaireMaxMinutes;
    }

    public int getDureeHebdomadaireMaxMineurMinutes() {
        return dureeHebdomadaireMaxMineurMinutes;
    }

    public void setDureeHebdomadaireMaxMineurMinutes(int dureeHebdomadaireMaxMineurMinutes) {
        this.dureeHebdomadaireMaxMineurMinutes = dureeHebdomadaireMaxMineurMinutes;
    }

    public int getDureeVacationMaxMinutes() {
        return dureeVacationMaxMinutes;
    }

    public void setDureeVacationMaxMinutes(int dureeVacationMaxMinutes) {
        this.dureeVacationMaxMinutes = dureeVacationMaxMinutes;
    }

    public int getReposQuotidienMinimalMinutes() {
        return reposQuotidienMinimalMinutes;
    }

    public void setReposQuotidienMinimalMinutes(int reposQuotidienMinimalMinutes) {
        this.reposQuotidienMinimalMinutes = reposQuotidienMinimalMinutes;
    }

    public int getDureePauseMinutes() {
        return dureePauseMinutes;
    }

    public void setDureePauseMinutes(int dureePauseMinutes) {
        this.dureePauseMinutes = dureePauseMinutes;
    }

    /**
     * The break owed to this animateur on that date: the edition's own
     * duration, raised to the thirty minutes of art. <b>L3162-3</b> for a young
     * worker. That floor is d'ordre public, so an edition that sets twenty-five
     * gives twenty-five to its adults and thirty to its minors rather than be
     * refused outright — and every reader of a break length, rules, eligibility
     * filter, Pauses screen and hours alike, goes through here, so none of them
     * can be stricter or laxer than the others (ADR 0048).
     */
    public int dureePauseMinutes(boolean mineur) {
        return mineur ? Math.max(dureePauseMinutes, PlafondsLegauxMineurs.PAUSE_MINIMALE_MINUTES) : dureePauseMinutes;
    }

    public int getCoupureRepasMinutes() {
        return coupureRepasMinutes;
    }

    public void setCoupureRepasMinutes(int coupureRepasMinutes) {
        this.coupureRepasMinutes = coupureRepasMinutes;
    }

    public LocalTime getCoupureRepasMidiDebut() {
        return coupureRepasMidiDebut;
    }

    public void setCoupureRepasMidiDebut(LocalTime coupureRepasMidiDebut) {
        this.coupureRepasMidiDebut = coupureRepasMidiDebut;
    }

    public LocalTime getCoupureRepasMidiFin() {
        return coupureRepasMidiFin;
    }

    public void setCoupureRepasMidiFin(LocalTime coupureRepasMidiFin) {
        this.coupureRepasMidiFin = coupureRepasMidiFin;
    }

    public LocalTime getCoupureRepasSoirDebut() {
        return coupureRepasSoirDebut;
    }

    public void setCoupureRepasSoirDebut(LocalTime coupureRepasSoirDebut) {
        this.coupureRepasSoirDebut = coupureRepasSoirDebut;
    }

    public LocalTime getCoupureRepasSoirFin() {
        return coupureRepasSoirFin;
    }

    public void setCoupureRepasSoirFin(LocalTime coupureRepasSoirFin) {
        this.coupureRepasSoirFin = coupureRepasSoirFin;
    }

    public LocalTime getHeureDebutSoiree() {
        return heureDebutSoiree;
    }

    public void setHeureDebutSoiree(LocalTime heureDebutSoiree) {
        this.heureDebutSoiree = heureDebutSoiree;
    }
}

File: src/main/webui/src/app/pages/parametres/parametres-legaux.ts
import { ChangeDetectionStrategy, Component, inject, signal } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { MatSlideToggleModule } from '@angular/material/slide-toggle';
import { RouterLink } from '@angular/router';
import { ConstraintsApi } from '../../core/api/constraints-api';
import { errorPrefix } from '../../core/error-message';
import { urlLegifrance } from '../../core/legifrance';
import {
  DUREE_HEBDOMADAIRE_MAX_HEURES,
  DUREE_HEBDOMADAIRE_MAX_MINEUR_HEURES,
  ParametresLegaux,
} from '../../core/models';

/**
 * The legal parameters of the edition — the two weekly ceilings, the daily
 * rest, the gap between vacations, the on-post break, the hour the evening
 * starts at for the Équité screen — and, as the card right
 * below, the meal break: one record, two cards of the Paramètres page, either
 * save button sending the whole record. They used to sit on the Contraintes
 * page, next to the rules that read them; that is where nobody looked for
 * them, so every parameter of the edition is now set in one place.
 */
@Component({
  selector: 'app-parametres-legaux',
  imports: [
    FormsModule,
    MatCardModule,
    MatButtonModule,
    MatIconModule,
    MatFormFieldModule,
    MatInputModule,
    MatSlideToggleModule,
    RouterLink,
  ],
  templateUrl: './parametres-legaux.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ParametresLegauxCard {
  protected readonly parametresLoading = signal(false);
  protected readonly parametresError = signal('');
  protected readonly parametresSaved = signal(false);
  protected readonly dureeHebdomadaireMaxHeures = signal<number | null>(null);
  protected readonly pauseSurPoste = signal(false);
  /** Minimum gap between two same-day vacations, in minutes (no legal floor: 0 lets blocks chain). */
  protected readonly gapBetweenVacationsMinutes = signal<number | null>(null);
  /** Minimum daily rest, in hours (art. L3131-1: 11 h; a lower value needs a collective agreement). */
  protected readonly reposQuotidienHeures = signal<number | null>(null);
  protected readonly articleReposQuotidien = urlLegifrance('L3131-1');
  protected readonly articlePause = urlLegifrance('L3121-16');
  protected readonly dureeHebdomadaireMaxMineurHeures = signal<number | null>(null);
  /** The meal break the organisation sets itself (issue #438): its length, and the two windows as HH:MM. */
  protected readonly coupureRepasMinutes = signal<number | null>(null);
  protected readonly coupureRepasMidiDebut = signal('');
  protected readonly coupureRepasMidiFin = signal('');
  protected readonly coupureRepasSoirDebut = signal('');
  protected readonly coupureRepasSoirFin = signal('');
  /** When the evening starts for the Équité screen (HH:MM), the organisation's rule (issue #497). */
  protected readonly heureDebutSoiree = signal('');
  /**
   * How long one vacation may run before the grid check warns. It used to bound
   * what the découpage produced; the découpage is gone and the rule it encoded
   * is not, so it is edited here with the other rules of the event.
   */
  protected readonly dureeVacationMaxHeures = signal<number | null>(null);

  /** Ordre public ceilings, mirrored from the server-side validation. */
  protected readonly ceilingAdultHours = DUREE_HEBDOMADAIRE_MAX_HEURES;
  protected readonly ceilingMinorHours = DUREE_HEBDOMADAIRE_MAX_MINEUR_HEURES;

  // The two ceilings are ordre public: the card states which article says so,
  // and links it, rather than asking the reader to take our word for it.
  protected readonly articleCeilingAdult = urlLegifrance('L3121-20');
  protected readonly articleCeilingMinor = urlLegifrance('L3162-1');

  private readonly constraintsApi = inject(ConstraintsApi);

  constructor() {
    void this.loadParametresLegaux();
  }

  protected async loadParametresLegaux(): Promise<void> {
    this.parametresLoading.set(true);
    this.parametresError.set('');
    try {
      this.read(await this.constraintsApi.legalParameters());
    } catch (error) {
      this.parametresError.set(errorPrefix(error));
    } finally {
      this.parametresLoading.set(false);
    }
  }

  private read(parametres: ParametresLegaux): void {
    this.dureeHebdomadaireMaxHeures.set(parametres.dureeHebdomadaireMaxMinutes / 60);
    this.dureeHebdomadaireMaxMineurHeures.set(parametres.dureeHebdomadaireMaxMineurMinutes / 60);
    this.gapBetweenVacationsMinutes.set(parametres.pauseMinimaleEntreVacationsMinutes);
    this.reposQuotidienHeures.set(parametres.reposQuotidienMinimalMinutes / 60);
    this.pauseSurPoste.set(parametres.pauseSurPoste);
    this.coupureRepasMinutes.set(parametres.coupureRepasMinutes);
    this.coupureRepasMidiDebut.set(parametres.coupureRepasMidiDebut ?? '');
    this.coupureRepasMidiFin.set(parametres.coupureRepasMidiFin ?? '');
    this.coupureRepasSoirDebut.set(parametres.coupureRepasSoirDebut ?? '');
    this.coupureRepasSoirFin.set(parametres.coupureRepasSoirFin ?? '');
    this.heureDebutSoiree.set(parametres.heureDebutSoiree ?? '');
    this.dureeVacationMaxHeures.set(parametres.dureeVacationMaxMinutes / 60);
  }

  /**
   * Saves every legal parameter the card holds — the two weekly ceilings, the
   * gap between vacations, the daily rest, the on-post break, the meal break
   * and the vacation ceiling — so that a save never silently resets a field the card did not
   * show: the server replaces the whole record. The bounds mirror the
   * server-side check (`ReferenceDataService.updateParametresLegaux`): a value
   * above the ordre public maximum is refused here too, so the administrator
   * gets an explanation rather than an HTTP 500. A lower value stays free — it
   * is more protective than the law.
   */
  protected async saveParametresLegaux(): Promise<void> {
    const heures = this.dureeHebdomadaireMaxHeures();
    const heuresMineur = this.dureeHebdomadaireMaxMineurHeures();
    const pauseMinutes = this.gapBetweenVacationsMinutes();
    const reposHeures = this.reposQuotidienHeures();
    const coupureMinutes = this.coupureRepasMinutes();
    const vacationMaxHeures = this.dureeVacationMaxHeures();
    if (
      heures === null ||
      heures <= 0 ||
      heuresMineur === null ||
      heuresMineur <= 0 ||
      pauseMinutes === null ||
      pauseMinutes < 0 ||
      reposHeures === null ||
      reposHeures < 0 ||
      coupureMinutes === null ||
      coupureMinutes < 0 ||
      vacationMaxHeures === null ||
      vacationMaxHeures <= 0 ||
      this.heureDebutSoiree() === ''
    ) {
      return;
    }
    if (heures > this.ceilingAdultHours) {
      this.parametresError.set(
        $localize`:@@constraints.legal.error.plafondMajeur:La durée hebdomadaire maximale des majeurs ne peut pas dépasser ${this.ceilingAdultHours}:hours: h (Code du travail art. L3121-20, disposition d'ordre public).`,
      );
      return;
    }
    if (heuresMineur > this.ceilingMinorHours) {
      this.parametresError.set(
        $localize`:@@constraints.legal.error.plafondMineur:La durée hebdomadaire maximale des mineurs ne peut pas dépasser ${this.ceilingMinorHours}:hours: h (Code du travail art. L3162-1).`,
      );
      return;
    }
    this.parametresLoading.set(true);
    this.parametresError.set('');
    this.parametresSaved.set(false);
    try {
      this.read(
        await this.constraintsApi.saveLegalParameters({
          dureeHebdomadaireMaxMinutes: Math.round(heures * 60),
          dureeHebdomadaireMaxMineurMinutes: Math.round(heuresMineur * 60),
          pauseMinimaleEntreVacationsMinutes: Math.round(pauseMinutes),
          reposQuotidienMinimalMinutes: Math.round(reposHeures * 60),
          pauseSurPoste: this.pauseSurPoste(),
          coupureRepasMinutes: Math.round(coupureMinutes),
          coupureRepasMidiDebut: this.coupureRepasMidiDebut(),
          coupureRepasMidiFin: this.coupureRepasMidiFin(),
          coupureRepasSoirDebut: this.coupureRepasSoirDebut(),
          coupureRepasSoirFin: this.coupureRepasSoirFin(),
          heureDebutSoiree: this.heureDebutSoiree(),
          dureeVacationMaxMinutes: Math.round(vacationMaxHeures * 60),
        }),
      );
      this.parametresSaved.set(true);
    } catch (error) {
      this.parametresError.set(errorPrefix(error));
    } finally {
      this.parametresLoading.set(false);
    }
  }
}
